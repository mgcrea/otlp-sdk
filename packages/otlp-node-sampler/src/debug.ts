import createdDebug from "debug";
export const debug = createdDebug("otlp-node-sampler");
