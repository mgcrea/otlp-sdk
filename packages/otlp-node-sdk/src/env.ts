export const NODE_ENV = process.env["NODE_ENV"];

export const TRACING_ENABLED =
  process.env["TRACING_ENABLED"] && process.env["TRACING_ENABLED"] !== "0";

export const TRACING_SERVICE_NAME = process.env["TRACING_SERVICE_NAME"];

export const TRACING_SERVICE_VERSION = process.env["TRACING_SERVICE_VERSION"];

export const TRACING_OTLP_URL = process.env["TRACING_OTLP_URL"];

export const TRACING_IP_ALLOWLIST = process.env["TRACING_IP_ALLOWLIST"]
  ? process.env["TRACING_IP_ALLOWLIST"].split(",")
  : [];

export const TRACING_USER_AGENT_DENYLIST = process.env["TRACING_USER_AGENT_DENYLIST"]
  ? process.env["TRACING_USER_AGENT_DENYLIST"].split("|")
  : [];
