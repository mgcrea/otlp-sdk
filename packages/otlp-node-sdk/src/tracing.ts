import { NodeSampler, type NodeSamplerOptions } from "@mgcrea/otlp-node-sampler";
import { DiagConsoleLogger, DiagLogLevel, diag } from "@opentelemetry/api";
import { getNodeAutoInstrumentations } from "@opentelemetry/auto-instrumentations-node";
import { OTLPMetricExporter } from "@opentelemetry/exporter-metrics-otlp-proto";
import { OTLPTraceExporter } from "@opentelemetry/exporter-trace-otlp-proto";
import { registerInstrumentations } from "@opentelemetry/instrumentation";
import { FastifyInstrumentation } from "@opentelemetry/instrumentation-fastify";
import {
  HttpInstrumentation,
  type HttpInstrumentationConfig,
} from "@opentelemetry/instrumentation-http";
import { Resource } from "@opentelemetry/resources";
import { BatchSpanProcessor, ConsoleSpanExporter } from "@opentelemetry/sdk-trace-base";
import { NodeTracerProvider } from "@opentelemetry/sdk-trace-node";
import { SemanticResourceAttributes } from "@opentelemetry/semantic-conventions";
import { PrismaInstrumentation } from "@prisma/instrumentation";
import { IncomingMessage } from "http";
import {
  NODE_ENV,
  TRACING_IP_ALLOWLIST,
  TRACING_SERVICE_NAME,
  TRACING_SERVICE_VERSION,
  TRACING_USER_AGENT_DENYLIST,
} from "./env";

const isNotNullish = <T>(input: T | null | undefined): input is T => input != null;

const defaultHttpRequestHook: HttpInstrumentationConfig["requestHook"] = (span, request) => {
  if (request instanceof IncomingMessage) {
    const { method, url } = request;
    span.updateName(`${method} ${url}`);
  }
};

type KnownConfigs = {
  node: Parameters<typeof getNodeAutoInstrumentations>[0];
  http: ConstructorParameters<typeof HttpInstrumentation>[0];
  fastify: ConstructorParameters<typeof FastifyInstrumentation>[0];
  prisma: ConstructorParameters<typeof PrismaInstrumentation>[0];
};

type ObjectToTupleUnion<T extends Record<string, unknown>> = {
  [K in keyof T]: [K, T[K]];
}[keyof T];

type KnownPlugins = keyof KnownConfigs;
type KnownPluginsWithOptions = ObjectToTupleUnion<KnownConfigs>;
// type KnownPluginOptions = ExtractSecondItem<KnownConfigs>;

type ConfigureTracingOptions = Pick<NodeSamplerOptions, "ipAllowList" | "userAgentDenyList"> & {
  name: string;
  version: string;
  env?: string;
  logLevel?: DiagLogLevel;
  plugins?: Array<KnownPlugins | KnownPluginsWithOptions>;
  traceExporterOptions?: ConstructorParameters<typeof OTLPTraceExporter>[0];
  metricExporterOptions?: ConstructorParameters<typeof OTLPMetricExporter>[0];
};

export const configureTracing = (options: ConfigureTracingOptions) => {
  const {
    logLevel = DiagLogLevel.INFO,
    name = TRACING_SERVICE_NAME,
    version = TRACING_SERVICE_VERSION,
    env = NODE_ENV,
    ipAllowList = TRACING_IP_ALLOWLIST,
    userAgentDenyList = TRACING_USER_AGENT_DENYLIST,
    plugins = [],
    traceExporterOptions = {},
    // metricExporterOptions = {},
  } = options;

  diag.setLogger(new DiagConsoleLogger(), logLevel);

  const resource = Resource.default().merge(
    new Resource({
      [SemanticResourceAttributes.SERVICE_NAME]: name,
      [SemanticResourceAttributes.SERVICE_VERSION]: version,
      [SemanticResourceAttributes.DEPLOYMENT_ENVIRONMENT]: env,
    })
  );

  const sampler = new NodeSampler({
    ipAllowList,
    userAgentDenyList,
  });

  const provider = new NodeTracerProvider({
    resource,
    sampler,
  });

  const config = plugins.reduce<Partial<KnownConfigs>>((soFar, value) => {
    const name = Array.isArray(value) ? value[0] : value;
    const options = Array.isArray(value) ? value[1] : {};
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    soFar[name] = options as any;
    return soFar;
  }, {});

  const instrumentations = [
    config.node ? getNodeAutoInstrumentations(config.node) : null,
    config.http
      ? new HttpInstrumentation({
          requestHook: defaultHttpRequestHook,
          ...config.http,
        })
      : null,
    config.fastify
      ? new FastifyInstrumentation({
          ...config.fastify,
        })
      : null,
    config.prisma
      ? new PrismaInstrumentation({
          ...config.prisma,
        })
      : null,
  ].filter(isNotNullish);

  registerInstrumentations({ instrumentations, tracerProvider: provider });

  const traceExporter = traceExporterOptions.url
    ? new OTLPTraceExporter(traceExporterOptions)
    : new ConsoleSpanExporter();
  const spanProcessor = new BatchSpanProcessor(traceExporter);
  provider.addSpanProcessor(spanProcessor);

  // @TODO
  // const metricExporter = metricExporterOptions.url
  //   ? new OTLPMetricExporter(metricExporterOptions)
  //   : new ConsoleMetricExporter();
  // const metricReader = new PeriodicExportingMetricReader({ exporter: metricExporter });

  // const consoleExporter = new ConsoleSpanExporter();
  // provider.addSpanProcessor(new SimpleSpanProcessor(consoleExporter));

  provider.register();

  // const sdk = new NodeSDK({
  //   resource,
  //   sampler,
  //   spanProcessor,
  //   metricReader,
  //   instrumentations,
  // });
  // sdk.start();
};
