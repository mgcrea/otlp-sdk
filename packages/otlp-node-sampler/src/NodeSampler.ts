/* eslint-disable testing-library/no-debugging-utils */
import { SpanKind, TraceFlags, type Attributes, type Context } from "@opentelemetry/api";
import { getSpan } from "@opentelemetry/api/build/src/trace/context-utils";
import { SamplingDecision, type Sampler, type SamplingResult } from "@opentelemetry/sdk-trace-base";
import { debug } from "./debug";

export type NodeSamplerOptions = {
  ipAllowList?: string[];
  userAgentDenylist?: string[];
  sampleByDefault?: boolean;
};

export class NodeSampler implements Sampler {
  private ipAllowList: Set<string>;
  private userAgentDenylist: Set<string>;
  private sampleByDefault: boolean;

  static RECORD_AND_SAMPLED: SamplingResult = {
    decision: SamplingDecision.RECORD_AND_SAMPLED,
    attributes: { ip_sampled: "yes" },
  };
  static NOT_RECORD: SamplingResult = {
    decision: SamplingDecision.NOT_RECORD,
  };

  constructor({ ipAllowList, userAgentDenylist, sampleByDefault }: NodeSamplerOptions = {}) {
    this.ipAllowList = new Set(ipAllowList);
    this.userAgentDenylist = new Set(userAgentDenylist);
    this.sampleByDefault = Boolean(sampleByDefault);
  }

  shouldSample(
    context: Context,
    _traceId: string,
    spanName: string,
    _spanKind: SpanKind,
    attributes: Attributes
  ): SamplingResult {
    const parentSpan = getSpan(context);

    // Inherit sampling decision from parent span
    if (parentSpan) {
      const parentIsSampled = parentSpan.spanContext().traceFlags === TraceFlags.SAMPLED;
      return parentIsSampled ? NodeSampler.RECORD_AND_SAMPLED : NodeSampler.NOT_RECORD;
    }

    // Filter by user agent
    if (this.userAgentDenylist.size > 0) {
      const userAgent = attributes["http.user_agent"]?.toString();
      const isDeniedUserAgent = userAgent && this.userAgentDenylist.has(userAgent);
      if (isDeniedUserAgent) {
        debug(`Ignored span ${spanName} with client ip ${userAgent}}`);
        return NodeSampler.NOT_RECORD;
      }
    }

    // Filter by ip address
    if (this.ipAllowList.size > 0) {
      const clientIp = attributes["http.client_ip"]?.toString();
      const isAllowedIp = clientIp && this.ipAllowList.has(clientIp);
      if (!isAllowedIp) {
        debug(`Ignored span ${spanName} with client ip ${clientIp}}`);
        return NodeSampler.NOT_RECORD;
      }
    }

    if (this.sampleByDefault) {
      return NodeSampler.RECORD_AND_SAMPLED;
    }
    debug(`Ignored span ${spanName} with attribute ${JSON.stringify(attributes)}`);
    return NodeSampler.NOT_RECORD;
  }

  toString(): string {
    return "NodeSampler";
  }
}
