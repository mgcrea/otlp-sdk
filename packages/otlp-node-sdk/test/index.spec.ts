import { describe, expect, it } from "vitest";
import * as module from "../src/index";

describe("module", () => {
  it("should export an eslint config", () => {
    expect(module).toMatchSnapshot();
  });
});
