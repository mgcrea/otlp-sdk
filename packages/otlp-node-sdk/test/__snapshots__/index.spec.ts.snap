// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`module > should export an eslint config 1`] = `
{
  "configureTracing": [Function],
}
`;
